sympy/solvers/solveset.py

```diff
@@ -2553,12 +2555,17 @@ def linsolve(system, *symbols):
                     symbols for which a solution is being sought must
                     be given as a sequence, too.
                 '''))
-            system = [
-                _mexpand(i.lhs - i.rhs if isinstance(i, Eq) else i,
-                recursive=True) for i in system]
-            system, symbols, swap = recast_to_symbols(system, symbols)
-            A, b = linear_eq_to_matrix(system, symbols)
-            syms_needed_msg = 'free symbols in the equations provided'
+            eqs = system
+            try:
+                eqs, ring = sympy_eqs_to_ring(eqs, symbols)
+            except PolynomialError as exc:
```

Comment:
Does this need to recognize ValueError?
```
File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/solvers/solveset.py", line 2525, in sympy.solvers.solveset.linsolve
Failed example:
    linsolve([x**2 - 1], x)
Expected:
    Traceback (most recent call last):
    ...
    NonlinearError:
    nonlinear term encountered: x**2
Got:
    Traceback (most recent call last):
      File "/opt/python/3.8.0/lib/python3.8/doctest.py", line 1328, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest sympy.solvers.solveset.linsolve[23]>", line 1, in <module>
        linsolve([x**2 - 1], x)
      File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/solvers/solveset.py", line 2564, in linsolve
        sol = solve_lin_sys(eqs, ring, _raw=False)
      File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/polys/solvers.py", line 73, in solve_lin_sys
        eq_coeffs = {ring.gens[monom.index(1)]:coeff for monom, coeff in eq_dict.items()}
      File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/polys/solvers.py", line 73, in <dictcomp>
        eq_coeffs = {ring.gens[monom.index(1)]:coeff for monom, coeff in eq_dict.items()}
    ValueError: tuple.index(x): x not in tuple
```

Reply:
I'm not sure. We have `NonlinearError` now which should probably be raised in this case.